# <span style="color:#0b486b"> FIT3182: Big Data Management and Processing (2025) </span>
---

Teaching Team:

Faculty of Information Technology, Monash University, Australia
* A/Prof. David Taniar (Chief Examiner) | david.taniar@monash.edu

School of Information Technology, Monash University, Malaysia
* Vishnu Monn (Unit Coordinator) | vishnu.monn@monash.edu
* Shageenderan Sapai | shageenderan.sapai@monash.edu
* Henry Quan Bi Pay | quan.pay@monash.edu
* Ruturaj Reddy | ruturaj.reddy@monash.edu
* Chai Wai Jin (Class Assistant) | wcha0106@student.monash.edu



---

# <span style="color:#0b486b">  Group Information</span>
---
Note: Group members need to be enrolled in the same tutorial day and time slot.

Your tutorial day and time: Thursday 6pm    <br/>

1st group member

Surname: Kang  <br/>
Firstname: Jia Xin    <br/>
Student ID: 33200556   <br/>
Email: jkan0037@student.monash.edu   <br/>

2nd group member

Surname: Lok  <br/>
Firstname: Mei Hui    <br/>
Student ID: 33458715    <br/>
Email: mlok0006@student.monash.edu    <br/>


---

# Streaming Application
### Due: <span style="color:red">11:55pm MYT, 27th May 2025</span>  (Tuesday)

#### <span style="color:red">Important note:</span> This is an **group** assignment with two students (max) per group. You or your group partner can share the code and outcomes of this assignment. However, you should not attempt to post questions on EdForum or any other online platform seeking solutions to the answers. If you require clarification on the assignment questions, you can post a post on EdForum or seek consultation from the tutors. In addition, AI and generative tools may be used in Guided ways.  However, students will be required to demonstrate a comprehensive understanding of the submitted work, failing which significant marks will be deducted from the submitted work. Even though this is a group work, each student is required to submit the assignment work in Moodle.

## Instructions
This notebook has been prepared for you to complete Assignment 2. The theme of this assignment is about practical knowledge and skills in streaming application using Spark and Kafka. **The total marks for this notebook is 30 marks, which is equivalent to 30 percentage points of the total coursework marks for this unit.**

* Before getting started, you should read the entire notebook carefully once to understand what you need to do.

* Always use the data from the provided `.csv` files to answer the questions unless stated otherwise.

This assignment contain **3 parts**:

* **Part 1**: MongoDB Data Model (5 Marks)
* **Part 2**: Streaming Application (20 Marks)
* **Part 3**: Documentation and comments to describe the proposed solution in the submitted notebook (5 Marks)
* **Part 4**: Code demo and interview (Negative marking)

Required Software:

* You will be using Python 3. Answer all questions inside this Jupyter Notebook
* Please use the provided Docker to load the Jupyter Notebook

**Hint**: This assignment was essentially designed based on the seminars and applied sessions covered from Week 6 to Week 11. You are strongly encouraged to go through these contents thoroughly which might help you to complete the assignment.

### Assignment Marking

The marking of this assignment is based on quality of work you have submitted rather than just quantity. Marking starts from 0 and goes up based on tasks you have successfully completed and their quality, for example, how well the code submitted follows programming standards, code documentation, presentation of the assignment, readability of the code, organization of the code and so on. Please find the PEP 8 -- Style Guide for Python Code [here](https://www.python.org/dev/peps/pep-0008/) for your reference. Please refer to marking guidelines in Moodle for further details.

### What to Submit

This assignment is to be completed individually and submitted to Moodle unit site. By the due date, you are required to submit the following files to the corresponding Assignment (Dropbox) in Moodle.

* **xxx_assignment02_data_design_streaming.ipynb**: this is your main Python notebook solution source file (the data design and streaming application).
* **xxx_assignment02_producer_a/b/c.ipynb**: this is your Python notebook solution to run the Kafka producer that reads from one of the camera event files. If you are running multiple producers concurrently in the main notebook, then this file is optional.
* **xxx_assignment02_visualisation.ipynb**: this is your Python notebook solution containing the data visualisation.
* **xxx_assignment02_code.zip** (if applicable): this is a zip file that contains python files with custom-defined classes and functions to be used in notebook.

where `xxx` represents the student ID of each group member. For example, if your student ID is <span style="color:red">12345</span> and your group partner's ID is is <span style="color:red">54321</span>, then your submission file name would be <span style="color:red">12345_54321_assignment02_data_design_streaming.ipynb</span>. Please do the same for all of the submission files.

Your assignment will be assessed based on the content of the submitted files in Moodle. We will use the same docker image as provided in this unit when marking your assignment. **If you used additional libraries, please include pip commands in your Jupyter notebook.**

### Plagiarism and Collusion

Plagiarism and collusion are serious academic offenses at Monash University. Students must not share their work with any student. Students should consult policy linked [here](https://www.monash.edu/students/academic/policies/academic-integrity) for more information. See also the video linked on the Moodle page under the Assignment block.

The submitted notebook files will be checked for collusion or plagiarism. Students suspected of colluding or plagiarising the assignment will be reported to the Student Conduct and Complaints Department for academic misconduct. Consequently, your grade for this unit will be withheld until the investigation is complete. 

### Generative AI usage

AI & Generative AI tools may be used in GUIDED ways within this assessment / task as per the guidelines provided.
 
In this task, AI can be used as specified for one or more parts of the assessment task as per the instructions.
You may use AI to help you learn how to solve the assignment.

Where used, AI must be used responsibly, clearly documented and appropriately acknowledged (see [Learn HQ](https://www.monash.edu/student-academic-success/build-digital-capabilities/create-online/acknowledging-the-use-of-generative-artificial-intelligence)).
 
Any work submitted for a mark must:
represent a sincere demonstration of your human efforts, skills and subject knowledge that you will be accountable for.
adhere to the guidelines for AI use set for the assessment task.
reflect the University’s commitment to academic integrity and ethical behaviour.
Inappropriate AI use and/or AI use without acknowledgement will be considered a breach of academic integrity.

### Late submissions 
Extensions and other individual alterations to the assessment regime will only be considered using the University’s [Special Consideration Policy](https://www.monash.edu/students/admin/exams/changes/special-consideration). There is a 10% penalty per day, including weekends, for late submission. Please note that short extensions are not allowed for group submissions.

## <span style="color:#0b486b">Preliminary</span>

### Scenario Background

Malaysia’s road network consistently ranks among the busiest and most accident‑prone in Southeast Asia. Federal roads alone account for a significant proportion of traffic incidents, particularly during peak travel periods and festive seasons, when speed limits of up to 90 km/h (and 110 km/h on expressways) are frequently exceeded in an effort to cover long distances quickly. Since 2012, the Automated Enforcement System (AES) has deployed static speed‑light and red‑light cameras at fixed points to deter speeding and dangerous cornering. However, these point‑capture devices suffer from well‑documented loopholes: drivers can simply decelerate when approaching a camera and then accelerate immediately afterward, rendering enforcement uneven and often ineffective.

To address these shortcomings, the Malaysian Government has begun rolling out the Automated Awareness Safety System (AWAS), a point‑to‑point average‑speed enforcement mechanism (Jamil et al., 2022). Figure 1 illustrates an overview of the AWAS system. AWAS leverages pairs of Ekin Spotter modular cameras equipped with 360° video surveillance and Automatic Number Plate Recognition (ANPR) to record each vehicle’s passage at two distinct checkpoints along a highway segment. By logging the exact timestamps at “Point A” and “Point B,” the system computes the travel time over a known distance (typically 1–5 km) and derives the average speed. Any average exceeding the legal limit (e.g., 110 km/h on expressways) automatically triggers a violation notice, regardless of momentary decelerations.

While AWAS promises more consistent enforcement, it also introduces significant data‑processing challenges. Each camera pair generates a continuous stream of high‑volume events—potentially thousands per minute during peak hours—that must be matched by license plate, ordered by event time, and joined across streams to compute speeds in near real time. The system must tolerate out‑of‑order or late‑arriving events (e.g., network delays), bound state growth via watermarks, and guarantee end‑to‑end exactly‑once processing to prevent duplicate violation records. These requirements make AWAS an ideal case study for a streaming Big Data architecture using Apache Kafka for ingestion, Apache Spark Structured Streaming for stateful stream–stream joins, and MongoDB for scalable storage of both raw events and flagged violations.

Reference:

Jamil, H. M., Shabadin, A., & Ibrahim, M. K. A. (2022). Automated Awareness Safety System (AwAS) for Red Light Running in Malaysia: An Analysis of Four-year Data on Its Effectiveness. Journal of the Society of Automotive Engineers Malaysia, 6(1), 19-29.

<div style="text-align: center">
    <img src="FIT3182_A2_Fig_1.png"></img>
    <p style="text-align: center">Figure 1 - Overview of AWAS</p>
</div>


### Dataset

In this assignment, you are provided the following `.csv` files to help you simulate the AWAS streaming application. The following details the information about the dataset.

#### vehicle.csv
* car_plate (a string-based unique identifier to each vehicle)
* owner_name (a string that contains name of the owner)
* owner_addr (a string that contains the address of the owner)
* vechicle_type (a string that represents the vechile model)
* registration_date (date and time when the vehicle was registered)

#### camera.csv
* camera_id (an integer-based unique identifier to camera location)
* latitude (a float value representing latitude of camera)
* longitude (a float value representing longitude of camera)
* position (a float value tells at which kilometer point is the camera)
* speed_limit (a float value of maximum legal speed for the segment)

#### camera_event.csv
* event_id (a string-based unique identifier to camera reading)
* batch_id (a integer-based identifier to batch reading)
* car_plate (a string-based unique identifier to each vehicle)
* camera_id (an integer-based unique identifier to camera location)
* timestamp (a string that tells the timestamp when the vehicle passed the camera)
* speed_reading (a float value that tells the instantaneous speed, recorded in km/h, by that camera)

#### camera_event_historic.csv
* violation_id (a string-based unique identifier for violation record)
* car_plate (a string-based unique identifier to each vehicle)
* camera_id_start (an integer-based unique identifier to starting camera location)
* camera_id_end (an integer-based unique identifier to ending camera location)
* timestamp_start (a string that tells the timestamp when the vehicle passed the starting camera)
* timestamp_end (a string that tells the timestamp when the vehicle passed the ending camera)
* speed_reading (a float value that tells the average speed, recorded in km/h, within the camera segment)

<span style="color:red">Important note:</span> Multiple files of camera_event.csv will be provided, each corresponds to a camera respectively.

### Required Imports

Import necessary Python modules in the cell below. Include `pip` statement if external libraries/modules are used.

In [1]:
# Add pip statement if necessary
!pip install pymongo
!pip install kafka-python --upgrade

In [2]:
# Add any imports here
from pymongo import MongoClient, ASCENDING, DESCENDING
from datetime import datetime
import pandas as pd
from json import loads
from pymongo import ReplaceOne
from bson.json_util import loads
from pymongo import UpdateOne, InsertOne, DeleteOne



from pyspark.sql import SparkSession
import os
from pyspark import SparkContext # spark
from pyspark.streaming import StreamingContext # spark streaming

## <span style="color:#0b486b">Part 1: MongoDB Data Model</span>

This section consists of 3 sub-questions

In this task, you will study the data model of a streaming application. You will demonstrate the theoretical knowledge by designing appropriate data model based on the provided dataset.

### Task 1.1 Collection Design

In this task, design **at least** the following 3 collections. Add other collections if they are necessary.
* Vehicle (Store static metadata about each vehicle)
* Camera (Store static definitions of each camera)
* Violation (Records of flagged violations)

For each collection, provide
* 1-2 sentence description of why this collection exists
* document schema and a sample document
* indexes (if any) by specifying
    * Fields (and sort order if applicable)
    * Type
    * Purpose of the index
* shard key strategy (if any) by specifying
    * Chosen shard key
    * Shard key type
    * Rationale
* data retention policy (if applicable)


---
### Vehicle Collection

**Purpose**  
Stores static metadata about each vehicle, such as its plate number, type and owner details. This allows enrichment of streaming violation data with contextual vehicle information.

**Schema:**
```json
{
  "_id": "ObjectId",
  "car_plate": "str",
  "owner_name": "str",
  "owner_addr": "str",
  "registration_date": "datetime",
  "vechicle_type": "str"
}
```
**Sample Document**
{'_id': ObjectId('683696c10ccdfcf8fbd6aac6'), 'car_plate': 'FT 02', 'owner_addr': '943 Jalan Bukit Mawar, Kuala Lumpur', 'owner_name': 'Goh Mei Wei', 'registration_date': datetime.datetime(2006, 8, 22, 3, 18), 'vechicle_type': 'Coupe'}

**Indexes**
* Field:car_plate 
* Type: Single Field 
* Purpose: Unique lookup during violation enrichment and stream joins.

**Shard Key Strategy**
none

**Data Retention Policy**
none

---
### Camera Collection

**Purpose**  
Stores fixed metadata for each AWAS or AES camera. Used for calculating distance and applicable speed limits between two camera points.

**Schema:**
```json
{
  "_id": "ObjectId",
  "camera_id": "float",
  "latitude": "float",
  "longitude": "float",
  "position": "float",
  "speed_limit": "float"
}
```

**Sample Document**
{'_id': ObjectId('683696c9f770ecd51494c0e7'), 'camera_id': 1.0, 'latitude': 2.157730731, 'longitude': 102.6601002, 'position': 152.5, 'speed_limit': 110.0}

**Indexes**
* Field: camera_id 
* Type: Single Field 
* Purpose: Fast lookup of camera details when joining with camera_event or calculating distance between cameras.

**Shard Key Strategy**
none

**Data Retention Policy**
none

---
### Violation Collection

**Purpose**  
Stores all violations identified by the AWAS system. Each document represents a completed point-to-point violation event.

**Schema:**
```json
{
  "_id": "ObjectId",
  "car_plate": "string",
  "violation_date": "datetime",
  "violation_record":[
    {
      "violation_id": "string",             
    "camera_id_start": "int",              
    "camera_id_end": "int",  
    "timestamp_start": "datetime",  
    "timestamp_end": "datetime",   
    "speed_reading": "double",          
    }
  ]
}
```

**Sample Document**
{'_id': ObjectId('6836b955d1ca29cc7dcdcfab'), 'car_plate': 'YE 6517', 'violation_date': datetime.datetime(2018, 11, 14, 0, 0), 'violation_records': [{'violation_id': '0ff38c74-7ee6-41cd-bc26-3a6060604a02', 'camera_id_start': 1, 'camera_id_end': 2, 'timestamp_start': datetime.datetime(2018, 11, 14, 8, 30, 11), 'timestamp_end': datetime.datetime(2018, 11, 14, 8, 30, 35), 'speed_reading': 145.0}]}


**Indexes**
* **Field:** `car_plate`  and `violation_date`
{ car_plate: 1, timestamp_start: -1 }
car plate in asceding, violation date in descending. 
* Type: compound index 
* **Purpose:** Efficiently retrieve recent violations for a specific car.

**Shard Key Strategy**
none

**Data Retention Policy**
none

In [3]:
# Connect to local MongoDB instance

# host_ip = "192.168.68.124"
# host_ip = "192.168.100.165"
host_ip = "10.192.0.245"
# 10.192.0.245

client = MongoClient(
    host=f'{host_ip}',
    port=27017
)

# Create the database
db = client["traffic_monitoring"]

In [4]:
df_vehicle = pd.read_csv("data/vehicle.csv")
df_camera = pd.read_csv("data/camera.csv")
df_violation = pd.read_csv("data/camera_event_historic.csv")

In [5]:
# testing
print(df_vehicle.head(3))

  car_plate           owner_name                              owner_addr  \
0     FT 02          Goh Mei Wei     943 Jalan Bukit Mawar, Kuala Lumpur   
1   DQZ 793  Muhammad bin Liyana  946 Jalan Bukit Jelutong, Kuala Lumpur   
2   GUT 393      Haziq bin Azlan   900 Jalan Bukit Bintang, Kuala Lumpur   

  vechicle_type    registration_date  
0         Coupe  2006-08-22T03:18:00  
1           SUV  2008-04-13T09:01:23  
2         Sedan  2008-01-20T20:50:45  


In [6]:
summary = df_vehicle.agg({
    "car_plate": ["nunique", "count"]
})

print(summary.rename(index={"nunique": "Unique Plates", "count": "Total Records"}))

               car_plate
Unique Plates       9844
Total Records      10000


In [7]:
db["vehicle"].drop() # same as db.vehicle.drop()
db["camera"].drop()
db["violation"].drop()

In [8]:
# vehicle index 
db["vehicle"].create_index("car_plate", unique=True) # ascending by default 

bulk_ops = []

for _, row in df_vehicle.iterrows():
    record = loads(row.to_json())
    reg_date = pd.to_datetime(record.pop("registration_date")).to_pydatetime()  # remove from record
    
    bulk_ops.append(UpdateOne(
        {"car_plate": record["car_plate"]},
        {
            "$setOnInsert": record,  # everything except registration_date
            "$max": {"registration_date": reg_date}
        },
        upsert=True
    ))

# Execute all updates in bulk
if bulk_ops:
    db.vehicle.bulk_write(bulk_ops)


In [9]:
for doc in db["vehicle"].find({}).limit(3):
    print(doc)
# index created 

{'_id': ObjectId('683696c10ccdfcf8fbd6aac6'), 'car_plate': 'FT 02', 'owner_addr': '943 Jalan Bukit Mawar, Kuala Lumpur', 'owner_name': 'Goh Mei Wei', 'registration_date': datetime.datetime(2006, 8, 22, 3, 18), 'vechicle_type': 'Coupe'}
{'_id': ObjectId('683696c10ccdfcf8fbd6aac7'), 'car_plate': 'DQZ 793', 'owner_addr': '946 Jalan Bukit Jelutong, Kuala Lumpur', 'owner_name': 'Muhammad bin Liyana', 'registration_date': datetime.datetime(2008, 4, 13, 9, 1, 23), 'vechicle_type': 'SUV'}
{'_id': ObjectId('683696c10ccdfcf8fbd6aac8'), 'car_plate': 'GUT 393', 'owner_addr': '900 Jalan Bukit Bintang, Kuala Lumpur', 'owner_name': 'Haziq bin Azlan', 'registration_date': datetime.datetime(2008, 1, 20, 20, 50, 45), 'vechicle_type': 'Sedan'}


In [10]:
# checking 
print(df_camera.head(3))

   camera_id  latitude   longitude  position  speed_limit
0          1  2.157731  102.660100     152.5          110
1          2  2.162419  102.652455     153.5          110
2          3  2.167353  102.644914     154.5           90


In [11]:
# camera index 
db["camera"].create_index("camera_id", unique=True)
# Prepare records
df_camera["camera_id"] = df_camera["camera_id"].astype(int)
df_camera["speed_limit"] = df_camera["speed_limit"].astype(int)

# Convert DataFrame rows to dicts for MongoDB
records = [loads(row.to_json()) for _, row in df_camera.iterrows()]

bulk_ops = [InsertOne(record) for record in records]

# Execute inserts
try:
    if bulk_ops:
        result = db.camera.bulk_write(bulk_ops, ordered=False)
        print(f" Inserted: {result.inserted_count} records")
except Exception as e:
    print(" Insert error:", e)

 Inserted: 3 records


In [12]:
for doc in db["camera"].find({}).limit(3):
    print(doc)
# index created 

{'_id': ObjectId('683696c9f770ecd51494c0e7'), 'camera_id': 1.0, 'latitude': 2.157730731, 'longitude': 102.6601002, 'position': 152.5, 'speed_limit': 110.0}
{'_id': ObjectId('683696c9f770ecd51494c0e8'), 'camera_id': 2.0, 'latitude': 2.162418757, 'longitude': 102.6524549, 'position': 153.5, 'speed_limit': 110.0}
{'_id': ObjectId('683696c9f770ecd51494c0e9'), 'camera_id': 3.0, 'latitude': 2.167352891, 'longitude': 102.6449144, 'position': 154.5, 'speed_limit': 90.0}


In [13]:
#checking
print(df_violation.head(3))

                           violation_id car_plate  camera_id_start  \
0  0ff38c74-7ee6-41cd-bc26-3a6060604a02   YE 6517                1   
1  2cf78fb9-bea2-4dd8-951a-0476057e9710     AN 73                2   
2  36136182-2be9-44d7-bebc-0a9b4d56b846      HC 0                1   

   camera_id_end             timestamp_start               timestamp_end  \
0              2         2018-11-14T08:30:11  2018-11-14T08:30:35.821118   
1              3  2018-08-30T08:29:18.936540  2018-08-30T08:29:51.037375   
2              2         2018-10-27T08:34:34  2018-10-27T08:35:04.346445   

   speed_reading  
0          145.0  
1          112.1  
2          118.6  


In [14]:
db["violation"].create_index(
    [("car_plate", 1), ("violation_date", -1)],
    unique=True
)
print(db["violation"].index_information())


{'_id_': {'v': 2, 'key': [('_id', 1)]}, 'car_plate_1_violation_date_-1': {'v': 2, 'key': [('car_plate', 1), ('violation_date', -1)], 'unique': True}}


In [15]:

result_df = df_violation[df_violation["car_plate"] == "PQY 9"]
print(result_df)
    

                               violation_id car_plate  camera_id_start  \
15391  4dbadeef-ef4b-4c68-9914-edc9faab40f5     PQY 9                2   
18027  b35aea26-538c-4599-88a1-8238a29af3cb     PQY 9                1   
23958  7e01ad0b-00b7-4ee3-9176-70caabb9e201     PQY 9                1   
29300  361dfca2-d271-480a-ae9c-925afbe50fd5     PQY 9                1   
38423  28b2a3d5-92b8-43ff-9e21-ea65c1a97226     PQY 9                2   
47918  848c848c-bdaf-43f9-9425-ff84da5200ac     PQY 9                2   

       camera_id_end             timestamp_start               timestamp_end  \
15391              3  2015-05-13T08:26:05.404447  2015-05-13T08:26:38.529249   
18027              2         2020-05-04T08:09:57  2020-05-04T08:10:21.156197   
23958              2         2018-12-09T08:18:37  2018-12-09T08:19:06.646131   
29300              2         2020-08-31T08:01:00  2020-08-31T08:01:29.422407   
38423              3  2018-10-11T08:32:46.902422  2018-10-11T08:33:24.141596   
4

In [16]:
inserted_count = 0
skipped_count = 0

for _, row in df_violation.iterrows():
    d_json = loads(row.to_json())

    car_plate = d_json["car_plate"].strip()
    timestamp_end = pd.to_datetime(d_json["timestamp_end"]).replace(microsecond=0, tzinfo=None)
    timestamp_start = pd.to_datetime(d_json["timestamp_start"]).replace(microsecond=0, tzinfo=None)
    violation_date = datetime.combine(timestamp_end.date(), datetime.min.time())

    query = {"car_plate": car_plate, "violation_date": violation_date}

    violation_record = {
        "violation_id": str(d_json["violation_id"]),
        "camera_id_start": int(d_json["camera_id_start"]),
        "camera_id_end": int(d_json["camera_id_end"]),
        "timestamp_start": timestamp_start,
        "timestamp_end": timestamp_end,
        "speed_reading": d_json["speed_reading"]
    }

    existing = db.violation.find_one(query)

    if existing is None:
        db.violation.insert_one({
            "car_plate": car_plate,
            "violation_date": violation_date,
            "violation_records": [violation_record]
        })
        inserted_count += 1
    else:
        if not any(
            r["timestamp_start"] == timestamp_start and r["timestamp_end"] == timestamp_end
            for r in existing["violation_records"]
        ):
            db.violation.update_one(
                {"_id": existing["_id"]},
                {"$push": {"violation_records": violation_record}}
            )
        skipped_count += 1

print("Done.")
print("Inserted:", inserted_count)
print("Skipped:", skipped_count)
print("Total documents in DB:", db.violation.count_documents({}))

Done.
Inserted: 50000
Skipped: 0
Total documents in DB: 50000


### Task 1.2 Collection Relationship

In this task, specify the relationships between collections and explain whether you choose to embed data or store references. Justify your choice in terms of:
* Read/write patterns
* Data duplication versus Join cost
* Consistency requirements (if applicable)

---

#### 1. Relationship: `violation.car_plate` → `vehicle.car_plates` (one to one because one car plate maps to one vehicle record.)

**Design Choice:** Store `car_plate` as a reference.

**Justification:**
- **Read/Write Pattern:** Typically, there is no direct read or write interaction between the Violation and Vehicle collections; the relationship exists solely as a reference.
- **Data Duplication vs Join Cost:** Storing a reference avoids duplication and reduces update propagation issues.
- **Consistency:** Referencing ensures data remains consistent even if vehicle details are updated separately.

---
   

#### 2. Relationship: `violation.violation_records[].camera_id_start and camera_id_end` → `camera.camera_id` (many to one, start camera and end camera may be the same camera)

**Design Choice:** Store `camera_id_start` and `camera_id_end` as a reference.


**Justification:**
- **Read/Write Pattern:** Typically, there is no direct read or write interaction between the violation and camera collections; the relationship exists solely as a reference.
- **Data Duplication vs Join Cost:** Camera data (e.g., geolocation, speed limit) is small but shared across many records. Referencing is more efficient and scalable.
- **Consistency:** Cameras are rarely updated. Referencing ensures data remains consistent even if vehicle details are updated separately.

---
   

#### 3. Relationship: `violation.violation_records[]` embedded in `violation ` (one to many, Each violation entry, grouped by date and car plate, contains one or more associated violation records.)

**Design Choice:** Embed the violation records within the violation document identified by the car plate and violation date.

**Justification:**
- **Read/Write Pattern:** Embedding is suitable because violations are typically written and queried in batches, grouped by car plate and date.
- **Data Duplication vs Join Cost:** The script inserts or updates daily violation documents per car, embedding unique records while preventing duplicates based on timestamps.
- **Consistency:** All violations for a specific car on a given day are stored in a single document, containing one or more unique violation records for that day.

---

### Task 1.3 Discussion

In this task, discuss whether your model supports
* Consistency and Idempotency
    * Does it support idempotent writes?
    * Explain any upsert pattern in `violation` collection
* Scalability and Fault-Tolerance
    * Can your data model support high ingest rates?
    * Can your data model support low-latency lookups?

Justify and explain the trade-off made in your design.

---

## 1. Consistency and Idempotency

### Idempotent Writes
Yes, the design supports idempotent writes. 
Idempotency is ensured by generating a unique violation_id via hashing car_plate, timestamp_start, and timestamp_end, and using $addToSet to prevent duplicates. update_one with upsert=True maintains one document per car per day, creating or updating as needed to group violations by date.

### Upsert Pattern in Violation Collection
In cases where a violation may need to be updated (e.g., speed re-evaluation or late-arriving timestamps), the system can perform an upsert operation.

Each violation record is uniquely defined by key fields and a deterministic violation_id, hashed from car_plate, timestamp_start, and timestamp_end. Using $addToSet prevents duplicates, while upsert=True ensures the document is created if absent or updated safely if it exists.


---
## 2. Scalability and Fault-Tolerance

### Support for High Ingest Rates
Yes, the data model is optimized for high write throughput. The model ingest high-throughput data via Apache Kafka and process it using Spark Structured Streaming with watermarks for event-time handling and late data.Data is processed in 5-second micro-batches to balance latency and throughput. For efficient writes, we use MongoDB’s bulk_write, reducing database round-trips and improving performance.

### Support for Low-Latency Lookups
Yes, query performance is optimized through indexing and referencing.
A compound index on car_plate and violation_date enables fast violation lookups, while a single-field index on camera_id speeds up camera detail queries. Embedding daily violations per car in one document further optimizes retrieval by eliminating joins.


---

## 3. Trade-offs and Justification

Vehicle and camera data are referenced rather than embedded, allowing global updates (e.g., speed limit changes) at the cost of additional lookups—mitigated by proper indexing. For writing, we use MongoDB’s bulk_write during the DbWriter.close() phase to reduce round-trips, trading off some local memory overhead, which is managed by limiting micro-batch intervals to 5 seconds.



## <span style="color:#0b486b">Part 2: Streaming Application</span>

This section consists of 2 sub-questions. 

In this task, you will implement a streaming application to simulate the AWAS system. Figure 2 illustrates an overview of the streaming architecture that is to be developed to simulate AWAS. Implementation is expected to be following programming standards with high readability (supported with documentation).

<div style="text-align: center"> 
    <img src="FIT3182_A2_Fig_2.png"></img>
    <p style="text-align: center">Figure 2 - Overview of streaming application to simulate AWAS </p>
</div>

### Task 2.1 Data Stream Processing

In this task, you will implement multiple **Apache Kafka** producers to simulate the real-time streaming of the data, which will be processed by **Apache Spark Structured Streaming** client and then inserted into MongoDB.

*<span style="color:red">Important note:</span> You are expected to use the same data model from Task 1. To make the streaming data consistent for the model, you may need to make some changes to the streaming data before building the model or inserting it to MongoDB.*

#### Event Producer

**Event Producer A**: Write a python program that loads all the data from `camera_event_A.csv` and feed the data to the stream in batches every $n$ seconds. You can refer to the batch id column in the csv file to identify the data to read and send to the stream. The typical value for $n$ is 5 seconds. You might need to append additional information such as producer information to identify the producer. If you are running this producer in a separate Jupyter notebook file, save the file as **xxx_assignment02_producer_a.ipynb**, where **xxx** represents the student IDs of the group members.

**Event Producer B**: Write a python program that loads all the data from `camera_event_B.csv` and feed the data to the stream in batches every $n$ seconds. You can refer to the batch id column in the csv file to identify the data to read and send to the stream. The typical value for $n$ is 5 seconds. You might need to append additional information such as producer information to identify the producer. If you are running this producer in a separate Jupyter notebook file, save the file as **xxx_assignment02_producer_b.ipynb**, where **xxx** represents the student IDs of the group members.

**Event Producer C**: Write a python program that loads all the data from `camera_event_C.csv` and feed the data to the stream in batches every $n$ seconds. You can refer to the batch id column in the csv file to identify the data to read and send to the stream. The typical value for $n$ is 5 seconds. You might need to append additional information such as producer information to identify the producer. If you are running this producer in a separate Jupyter notebook file, save the file as **xxx_assignment02_producer_c.ipynb**, where **xxx** represents the student IDs of the group members.

#### Streaming Application

Write a streaming application using Apache Spark Structured Streaming API which processes data in batches. Each batch should contain 0 or more camera event (from event producer). The streaming application should process the data as follows.
* Join the streaming data from the producers and determine if a vehicle should be flagged as violation. You should drop any data pair if the timestamp and the order of the camera does not match.
* If there is a violation detected, store it into MongoDB straight away.
* If there is no violation detected, drop the record.
* Due to the dynamic nature of moving vehicle, the time of vehicle completing the camera segment may vary and you should decide how many records and how long the records should be stored in the buffer until a pair is identified.

##### Violation Detection Rule
A vehicle is flagged as violating the speed limit if any one of the following happens.
* Instantaneous speed of vehicle exceed the speed limit of the recording camera
* Average speed of vehicle exceed the speed limit of the ending camera

<span style="color:red">Important Note:</span> *Only one record for a car per day is recorded in the database. If the car violates at **different cameras**, the record should be merged together into a single record to be stored in the database.*

## Producers

Each producer simulates a camera stream by sending batches of vehicle events from its respective CSV file. The producers operate on different batch intervals to emulate real-world asynchronous camera behavior.

- **Camera A Producer**: Sends a new batch every **10 seconds**
- **Camera B Producer**: Sends a new batch every **1 second**
- **Camera C Producer**: Sends a new batch every **1 second**

These producers publish events to the Kafka topics:
- `camera_event`

Explanation for time internal: 
internals of timestamps between batches are different across diff producer, thus we need to set different inetrval to acommodate for this, so that the batches will line up more closely, the watermark used in spark streaming will not eliminate certain data before it is able to join. 



## Stream Preprocessing

Each stream is assigned an alias and its column names are **renamed with numeric suffixes** (`_1`, `_2`, `_3`) to:
- Ensure clarity during joins
- Prevent column name conflicts
- Enable better traceability in logs and output

For example, `event_time` from Camera A becomes `event_time_1`, and so on.


## Speed Violation Detection Logic

We detect two types of violations:

#### 1. **Single-Point Violations (A,B,C)**  
Each stream is individually filtered to detect cases where a vehicle's `speed_reading` exceeds the legal threshold. These are simple violations where only one camera's data is used.
After detecting individual speed violations from **Camera A**, **Camera B**, and **Camera C**, each stream is standardized into a common format using a helper function. This ensures all streams share the same schema, with fields such as:

- `producer_id`
- `event_time`
- `event_id`
- `batch_id`
- `car_plate`
- `speed_reading`

we then **combine all single-point violations into one unified stream** using PySpark’s `union()` operation

---
#### 2. **Join-Based Violations (AB & BC)**  
We perform inner joins between:
- **Camera A & B**: (`AB_violation`)
- **Camera B & C**: (`BC_violation`)

**Join conditions:**
- Matching `car_plate`
- The second camera’s event time occurs *after* the first
- The time difference is *within 1 hour*

We calculate the **average speed** between the two cameras and flag it as a violation if it exceeds the limit.


## Stream-to-MongoDB Writers

There are three separate **Spark Structured Streaming writers**, each responsible for writing a different type of violation to MongoDB in real-time:

1. **AB Join Violation Writer**
   - Handles violations detected between Camera A and Camera B
   - Uses average speed between two event times and two camera locations
   - Calls `write_union_batch_to_mongo(df, batch_id, 1)` with violation type `1`

2. **BC Join Violation Writer**
   - Handles violations between Camera B and Camera C using the same join-based logic
   - Also uses violation type `1` since it’s a multi-camera event

3. **Single-Point Violation Writer**
   - Handles individual violations from Camera A, B, or C independently
   - Combines all single-point violations into one stream via `violations_union`
   - Uses violation type `2` to indicate it's a single-camera event

Each writer uses `foreachBatch()` to process micro-batches and write them to MongoDB with **exactly-once semantics**.


In [133]:
# import 
import os
import json
import uuid
from datetime import datetime
from pymongo import MongoClient
from pyspark.sql import SparkSession
from pyspark.sql.functions import from_json, col, to_timestamp, unix_timestamp, expr, explode
from pyspark.sql.types import StructType, StructField, StringType, IntegerType, DoubleType, TimestampType, ArrayType


In [134]:
"""
This section configures and initializes the Spark environment for reading Kafka streams.
It sets required packages, creates a local Spark session and adjusts time parsing settings for legacy compatibility.
"""

# Set up Spark environment for Kafka
os.environ['PYSPARK_SUBMIT_ARGS'] = '--packages org.apache.spark:spark-sql-kafka-0-10_2.12:3.3.0 pyspark-shell'

# Initialize Spark
spark = SparkSession.builder \
    .appName("Traffic Violation Detector") \
    .master("local[*]") \
    .getOrCreate()

spark.conf.set("spark.sql.legacy.timeParserPolicy", "LEGACY")

In [135]:
"""
This section sets the Kafka broker IP and topic name, then defines the schema used 
to parse the JSON messages from Kafka into structured Spark DataFrames.
"""

# Kafka configuration
# host_ip = "10.192.1.110"
# host_ip = "192.168.100.165"
host_ip = "10.192.0.245"
topic = "camera_events"

# Define schema for Kafka values
record = StructType([
    StructField("timestamp", StringType()),
    StructField("producer_id", IntegerType()),
    StructField("event_id", StringType()),
    StructField("batch_id", IntegerType()),
    StructField("camera_id", IntegerType()),
    StructField("car_plate", StringType()),
    StructField("speed_reading", DoubleType())
])

array = ArrayType(record)


In [136]:
"""
This section reads streaming data from the configured Kafka topic, 
parses the nested JSON structure and extracts relevant fields into a structured DataFrame.
"""
# Read Kafka stream
kafka_s = spark.readStream \
    .format("kafka") \
    .option("kafka.bootstrap.servers", f"{host_ip}:9092") \
    .option("subscribe", topic) \
    .load()

# Parse Kafka messages
parsed_s = kafka_s \
    .withColumn("json_str", col("value").cast("string")) \
    .withColumn("records", from_json(col("json_str"), array)) \
    .withColumn("record", explode(col("records"))) \
    .withColumn("event_time", to_timestamp(col("record.timestamp"), "yyyy-MM-dd'T'HH:mm:ss")) \
    .select(
        col("event_time"),
        col("record.producer_id"),
        col("record.event_id"),
        col("record.batch_id"),
        col("record.camera_id"),
        col("record.car_plate"),
        col("record.speed_reading"),
    )

In [137]:
def debug_batch(df, batch_id):
    print(f"\nBatch {batch_id} received at {datetime.utcnow().isoformat()}")
    count = df.count()
    if count == 0:
        print("Empty batch.")
    else:
        print(f"Rows in batch: {count}")
        df.show(truncate=False)


In [138]:
"""
This section connects to the MongoDB server and retrieves camera configuration details 
, like speed limits from the 'camera' collection in the 'traffic_monitoring' database.
The result is stored in a dictionary keyed by camera_id for quick access during stream processing.
"""

# MongoDB configuration and fetch camera info
mongo_client = MongoClient(host=host_ip, port=27017)
cam_s = {cam["camera_id"]: cam for cam in mongo_client["traffic_monitoring"]["camera"].find()}


In [139]:
"""
This section splits the parsed stream into three separate streams—one for each camera—
based on the `producer_id` field. A watermark of 90 minutes is applied to each stream to manage 
late data and enable proper event-time-based operations like joins and windowing.
"""

# Define Streams per producer_id 
stream_A = parsed_s.filter(col("producer_id") == "1").withWatermark("event_time", "90 minutes")
stream_B = parsed_s.filter(col("producer_id") == "2").withWatermark("event_time", "90 minutes")
stream_C = parsed_s.filter(col("producer_id") == "3").withWatermark("event_time", "90 minutes")


In [140]:
# Fetch Speed Limits 
A_speed = cam_s[1]["speed_limit"]
B_speed = cam_s[2]["speed_limit"]
C_speed = cam_s[3]["speed_limit"]


In [141]:
"""
This section renames key fields in each camera stream with numeric suffixes (_1, _2, _3) 
to indicate their origin (Camera A, B, or C). This naming convention ensures clear traceability 
and prevents column conflicts during join operations between streams.
Each renamed stream is also aliased ('a', 'b', 'c') for easier referencing in joins.
"""

# Rename Columns 
A = stream_A.selectExpr(
    "event_time as event_time_1",
    "event_id as event_id_1",
    "batch_id as batch_id_1",
    "car_plate as car_plate_1",
    "speed_reading as speed_1",
    "producer_id as producer_id_1"
).alias("a")

B = stream_B.selectExpr(
    "event_time as event_time_2",
    "event_id as event_id_2",
    "batch_id as batch_id_2",
    "car_plate as car_plate_2",
    "speed_reading as speed_2",
    "producer_id as producer_id_2"
).alias("b")

C = stream_C.selectExpr(
    "event_time as event_time_3",
    "event_id as event_id_3",
    "batch_id as batch_id_3",
    "car_plate as car_plate_3",
    "speed_reading as speed_3",
    "producer_id as producer_id_3"
).alias("c")



In [142]:
A.printSchema()


root
 |-- event_time_1: timestamp (nullable = true)
 |-- event_id_1: string (nullable = true)
 |-- batch_id_1: integer (nullable = true)
 |-- car_plate_1: string (nullable = true)
 |-- speed_1: double (nullable = true)
 |-- producer_id_1: integer (nullable = true)



In [143]:
"""
This section filters each camera stream (A, B and C) to detect single-point speed violations
where a vehicle exceeds the predefined speed limit. Each violation stream is then standardized
to a common schema using a helper function, making them compatible for union and downstream processing.

- `standardize_violation()` extracts and renames fields consistently.
- Standardized DataFrames (`A_std`, `B_std`, `C_std`) are then combined using `union()`
  to form a single unified stream, `violations_union`, containing all single-camera violations.
"""
# Single Camera Speed Violations
A_violate = A.filter(col("speed_1") > A_speed)
B_violate = B.filter(col("speed_2") > B_speed)
C_violate = C.filter(col("speed_3") > C_speed)

def standardize_violation(df, suffix):
    """
    Function that standardizes the column names of a violation DataFrame to a common schema
    so that it can be merged with violations from other cameras.
    """
    return df.selectExpr(
        f"producer_id_{suffix} as producer_id",
        f"event_time_{suffix} as event_time",
        f"event_id_{suffix} as event_id",
        f"batch_id_{suffix} as batch_id",
        f"car_plate_{suffix} as car_plate",
        f"speed_{suffix} as speed_reading"
    )

# Apply Standardization
A_std = standardize_violation(A_violate, "1")
B_std = standardize_violation(B_violate, "2")
C_std = standardize_violation(C_violate, "3")

# Union of single-point violations
violations_union = A_std.union(B_std).union(C_std)


In [144]:
rm -rf ./chk_violation_A

In [145]:
# testing 
a = A_std.writeStream \
    .outputMode("append") \
    .foreachBatch(debug_batch) \
    .option("checkpointLocation", "./chk_violation_A") \
    .start()


try:
    a.awaitTermination()
except KeyboardInterrupt:
    print("Stopping all ")
except Exception as e:
    print(e)
finally:
    a.stop()



Batch 0 received at 2025-05-28T07:04:26.655546
Empty batch.

Batch 1 received at 2025-05-28T07:04:32.307287
Rows in batch: 11
+-----------+-------------------+------------------------------------+--------+---------+-------------+
|producer_id|event_time         |event_id                            |batch_id|car_plate|speed_reading|
+-----------+-------------------+------------------------------------+--------+---------+-------------+
|1          |2024-01-01 08:00:01|d0e547bb-c4a7-4750-b7b4-8076e9b47f4f|1       |CJW 924  |148.3        |
|1          |2024-01-01 08:00:02|f3162606-1b2e-407f-951d-61d14c0a7b09|1       |CJP 278  |125.2        |
|1          |2024-01-01 08:00:00|5113d990-4da6-41d2-b292-2516e7d8dc07|1       |KKE 15   |150.3        |
|1          |2024-01-01 08:00:02|5a4497c9-5e3b-4470-a20c-8ee1f619f732|1       |CZ 592   |154.9        |
|1          |2024-01-01 08:00:00|b57e8e44-b6e5-4c1e-bb5b-a4e1bd24bac9|1       |UTT 229  |140.9        |
|1          |2024-01-01 08:00:04|c6965fdc

ERROR:root:KeyboardInterrupt while sending command.
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/site-packages/py4j/java_gateway.py", line 1038, in send_command
    response = connection.send_command(command)
  File "/opt/conda/lib/python3.8/site-packages/py4j/clientserver.py", line 511, in send_command
    answer = smart_decode(self.stream.readline()[:-1])
  File "/opt/conda/lib/python3.8/socket.py", line 669, in readinto
    return self._sock.recv_into(b)
KeyboardInterrupt


Stopping all 


In [146]:
rm -rf ./debug_batch

In [147]:
# testing 
u = violations_union.writeStream \
    .outputMode("append") \
    .foreachBatch(debug_batch) \
    .option("checkpointLocation", "./debug_batch") \
    .start()

try:
    u.awaitTermination()
except KeyboardInterrupt:
    print("Stopping all ")
except Exception as e:
    print(e)
finally:
    u.stop()


Batch 0 received at 2025-05-28T07:04:54.864124
Empty batch.

Batch 1 received at 2025-05-28T07:04:58.154252
Rows in batch: 11
+-----------+-------------------+------------------------------------+--------+---------+-------------+
|producer_id|event_time         |event_id                            |batch_id|car_plate|speed_reading|
+-----------+-------------------+------------------------------------+--------+---------+-------------+
|1          |2024-01-01 08:00:01|d0e547bb-c4a7-4750-b7b4-8076e9b47f4f|1       |CJW 924  |148.3        |
|1          |2024-01-01 08:00:02|f3162606-1b2e-407f-951d-61d14c0a7b09|1       |CJP 278  |125.2        |
|1          |2024-01-01 08:00:00|5113d990-4da6-41d2-b292-2516e7d8dc07|1       |KKE 15   |150.3        |
|1          |2024-01-01 08:00:02|5a4497c9-5e3b-4470-a20c-8ee1f619f732|1       |CZ 592   |154.9        |
|1          |2024-01-01 08:00:00|b57e8e44-b6e5-4c1e-bb5b-a4e1bd24bac9|1       |UTT 229  |140.9        |
|1          |2024-01-01 08:00:04|c6965fdc

Rows in batch: 14
+-----------+-------------------+------------------------------------+--------+---------+-------------+
|producer_id|event_time         |event_id                            |batch_id|car_plate|speed_reading|
+-----------+-------------------+------------------------------------+--------+---------+-------------+
|1          |2024-01-01 08:25:21|7399dc47-301e-4fef-83b5-0fb99c533df8|5       |IMU 122  |135.2        |
|1          |2024-01-01 08:25:21|654727a6-0002-46dd-ab19-8e468c001848|5       |GPR 4    |149.3        |
|1          |2024-01-01 08:25:22|42132abc-2c2e-4624-ab61-0769297f177b|5       |OW 5     |137.3        |
|1          |2024-01-01 08:25:21|e408f4bc-a7df-49bf-9523-1f5af6ec4627|5       |REP 98   |129.9        |
|1          |2024-01-01 08:25:20|97638b79-22e5-4d9f-9a37-12618005014b|5       |AFS 0    |134.1        |
|1          |2024-01-01 08:25:22|f302f720-927b-4c9e-b446-63000d364de8|5       |ZLY 15   |148.6        |
|1          |2024-01-01 08:25:21|eb604110-7b16

ERROR:root:KeyboardInterrupt while sending command.
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/site-packages/py4j/java_gateway.py", line 1038, in send_command
    response = connection.send_command(command)
  File "/opt/conda/lib/python3.8/site-packages/py4j/clientserver.py", line 511, in send_command
    answer = smart_decode(self.stream.readline()[:-1])
  File "/opt/conda/lib/python3.8/socket.py", line 669, in readinto
    return self._sock.recv_into(b)
KeyboardInterrupt
ERROR:py4j.clientserver:There was an exception while executing the Python Proxy on the Python Side.
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/site-packages/py4j/clientserver.py", line 617, in _call_proxy
    return_value = getattr(self.pool[obj_id], method)(*params)
  File "/opt/conda/lib/python3.8/site-packages/pyspark/sql/utils.py", line 272, in call
    raise e
  File "/opt/conda/lib/python3.8/site-packages/pyspark/sql/utils.py", line 269, in call
    self.func(Da

Stopping all 


In [148]:
"""
This section performs inter-camera joins to detect speed violations that occur 
between two different cameras—specifically, between Camera A & B (AB) and Camera B & C (BC).

## AB Join
- Joins records from Camera A and B based on matching `car_plate`.
- Ensures Camera B’s event time is after A’s and within 1 hour.
- Calculates the average speed between `event_time_1` and `event_time_2`.
- Filters out violations where the calculated `avg_speed` exceeds the speed limit 
  retrieved from MongoDB for Camera B. 

## BC Join
- Applies the same logic between Camera B and C.
- Ensures event order and time window between Camera B and C.
- Filters out violations where average speed exceeds the limit defined for Camera C. 

This method captures potential speeding events across multiple camera checkpoints 
by calculating travel speed over time, making it more reliable than single-point detection.
"""

# AB join
AB_violation = A.join(
    B,
    (col("a.car_plate_1") == col("b.car_plate_2")) &
    (col("b.event_time_2") > col("a.event_time_1")) &
    (col("b.event_time_2") <= col("a.event_time_1") + expr("INTERVAL 1 HOUR")),
    how="inner"
).selectExpr(
    "a.event_id_1", "a.batch_id_1", "a.speed_1", "a.event_time_1", "a.producer_id_1",
    "b.event_id_2", "b.batch_id_2", "b.speed_2", "b.event_time_2", "b.producer_id_2",
    "a.car_plate_1 as car_plate"
).withColumn(
    "avg_speed",
    1 / ((unix_timestamp(col("b.event_time_2")) - unix_timestamp(col("a.event_time_1"))) / 3600)
).filter(
    col("avg_speed") > cam_s[2]['speed_limit']
)



# BC join
BC_violation = B.join(
    C,
    (col("b.car_plate_2") == col("c.car_plate_3")) &
    (col("c.event_time_3") > col("b.event_time_2")) &
    (col("c.event_time_3") <= col("b.event_time_2") + expr("INTERVAL 1 HOUR")),
    how="inner"
).selectExpr(
    "b.event_id_2", "b.batch_id_2", "b.speed_2", "b.event_time_2", "b.producer_id_2",
    "c.event_id_3", "c.batch_id_3", "c.speed_3", "c.event_time_3", "c.producer_id_3",
    "b.car_plate_2 as car_plate"
).withColumn(
    "avg_speed",
    1 / ((unix_timestamp(col("c.event_time_3")) - unix_timestamp(col("b.event_time_2"))) / 3600)
).filter(
    col("avg_speed") > cam_s[3]['speed_limit']
)


In [149]:
AB_violate.printSchema()

root
 |-- A_event_time: timestamp (nullable = true)
 |-- A_event_id: string (nullable = true)
 |-- A_batch_id: integer (nullable = true)
 |-- A_car_plate: string (nullable = true)
 |-- A_speed: double (nullable = true)
 |-- A_producer_id: integer (nullable = true)
 |-- B_event_time: timestamp (nullable = true)
 |-- B_event_id: string (nullable = true)
 |-- B_batch_id: integer (nullable = true)
 |-- B_car_plate: string (nullable = true)
 |-- B_speed: double (nullable = true)
 |-- B_producer_id: integer (nullable = true)
 |-- avg_speed_h: double (nullable = true)



In [150]:
def debug_join_batch(df, batch_id):
    print("\n" + "=" * 90)
    print(f"Join VIOLATION DEBUG | Batch ID: {batch_id} | Logged at: {datetime.utcnow().isoformat()}")

    record_count = df.count()
    if record_count == 0:
        print("No violations detected in this batch.")
    else:
        print(f"Violations Detected: {record_count}")
        print("\nSample of Joined Violation Records:\n")
        df.show(truncate=False, n=20)

    print("=" * 90)


In [151]:
rm -rf ./ab_join_violate

In [152]:
ab = AB_violation.writeStream \
    .foreachBatch(debug_join_batch) \
    .outputMode("append") \
    .option("checkpointLocation", "./ab_join_violate") \
    .start()

try:
    ab.awaitTermination()
except KeyboardInterrupt:
    print("Stopping all ")
except Exception as e:
    print(e)
finally:
    ab.stop()


Join VIOLATION DEBUG | Batch ID: 0 | Logged at: 2025-05-28T07:05:58.203746
No violations detected in this batch.

Join VIOLATION DEBUG | Batch ID: 1 | Logged at: 2025-05-28T07:06:06.689579
No violations detected in this batch.

Join VIOLATION DEBUG | Batch ID: 2 | Logged at: 2025-05-28T07:06:13.174224
No violations detected in this batch.

Join VIOLATION DEBUG | Batch ID: 3 | Logged at: 2025-05-28T07:06:19.931652
No violations detected in this batch.

Join VIOLATION DEBUG | Batch ID: 4 | Logged at: 2025-05-28T07:06:26.895016
No violations detected in this batch.

Join VIOLATION DEBUG | Batch ID: 5 | Logged at: 2025-05-28T07:06:33.713204
Violations Detected: 1

Sample of Joined Violation Records:

+------------------------------------+----------+-------+-------------------+-------------+------------------------------------+----------+-------+-------------------+-------------+---------+------------------+
|event_id_1                          |batch_id_1|speed_1|event_time_1       |produ

ERROR:root:KeyboardInterrupt while sending command.
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/site-packages/py4j/java_gateway.py", line 1038, in send_command
    response = connection.send_command(command)
  File "/opt/conda/lib/python3.8/site-packages/py4j/clientserver.py", line 511, in send_command
    answer = smart_decode(self.stream.readline()[:-1])
  File "/opt/conda/lib/python3.8/socket.py", line 669, in readinto
    return self._sock.recv_into(b)
KeyboardInterrupt
ERROR:py4j.clientserver:There was an exception while executing the Python Proxy on the Python Side.
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/site-packages/py4j/clientserver.py", line 617, in _call_proxy
    return_value = getattr(self.pool[obj_id], method)(*params)
  File "/opt/conda/lib/python3.8/site-packages/pyspark/sql/utils.py", line 272, in call
    raise e
  File "/opt/conda/lib/python3.8/site-packages/pyspark/sql/utils.py", line 269, in call
    self.func(Da

Stopping all 


In [153]:
def write_union_batch_to_mongo(df, batch_id, violation_type):
    """
    This function writes a batch of traffic violation records to MongoDB. It supports both:
    - Join-based violations (AB, BC) → `violation_type = 1`
    - Single-point violations (A, B, C) → `violation_type != 1`

    Key features:
    - Generates a unique `violation_id` per record using car plate and timestamps

    - MongoDB checks if a document with the same `car_plate` and `violation_date` already exists.
    - If it does, it does not create a new document.
    - Instead, it adds the new `violation_record` to the `violation_records` array field using `$addToSet`.
    - `$addToSet` ensures that the same violation is not inserted twice, avoiding duplicates even if the stream retries.

    - If the document does not exist:
    - MongoDB will create a new document with the fields:
      - `car_plate`
      - `violation_date`
      - `violation_records`: initialized with the new violation entry
    - This is handled by setting `upsert=True`, which means "insert if not found".

    """
    print(f"\nWriting batch {batch_id} (type {violation_type}) to MongoDB")

    if df.isEmpty():
        print("Empty batch.")
        return

    try:
        mongo_client = MongoClient("10.192.0.245", 27017)
        mongo_client.admin.command("ping")
        db = mongo_client["traffic_monitoring"]
        violation_coll = db["violation"]
    except Exception as e:
        print(f" MongoDB connection failed: {e}")
        return

    bulk_ops = []

    for row in df.collect():
        try:
            data = row.asDict()
            print(data)
            car_plate = data["car_plate"]

            if violation_type == 1:
                # AB or BC violation (multi-camera)
                timestamp_start = data["event_time_1"]
                timestamp_end = data["event_time_2"]
                speed = data["avg_speed"]
                camera_id_start = int(data["producer_id_1"])
                camera_id_end = int(data["producer_id_2"])
            else:
                # Single-point violation (A, B, or C)
                timestamp_start = data["event_time"]
                timestamp_end = data["event_time"]
                speed = data["speed_reading"]
                camera_id_start = int(data["producer_id"])
                camera_id_end = camera_id_start

            violation_date = datetime.combine(timestamp_start.date(), datetime.min.time())
            violation_id = f"{car_plate}--{timestamp_end}--{timestamp_start}"

            violation_record = {
                "violation_id": violation_id,
                "camera_id_start": camera_id_start,
                "camera_id_end": camera_id_end,
                "timestamp_start": timestamp_start,
                "timestamp_end": timestamp_end,
                "speed_reading": speed
            }

            bulk_ops.append(
                UpdateOne(
                    {"car_plate": car_plate, "violation_date": violation_date},
                    {"$addToSet": {"violation_records": violation_record}},
                    upsert=True
                )
            )
        except Exception as e:
            print(f"Row error: {e}")

    try:
        if bulk_ops:
            violation_coll.bulk_write(bulk_ops, ordered=False)
            print(f"Wrote {len(bulk_ops)} violations to MongoDB.")
    except Exception as e:
        print(f" Bulk write error: {e}")
    finally:
        mongo_client.close()


In [154]:
rm -rf ./chk_violation_union_writer

In [ ]:
# testing 
violation_coll = db["violation"]
before_count = violation_coll.count_documents({})
print(f"Violation count BEFORE stream: {before_count}")

single = single = violations_union.writeStream \
    .foreachBatch(lambda df, batch_id: write_union_batch_to_mongo(df, batch_id, 2)) \
    .outputMode("append") \
    .option("checkpointLocation", "./chk_violation_union_writer") \
    .start()

try:
    single.awaitTermination()
except KeyboardInterrupt:
    print("Stopping all ")
except Exception as e:
    print(e)
finally:
    single.stop()

Violation count BEFORE stream: 50000

Writing batch 0 (type 2) to MongoDB
Empty batch.

Writing batch 1 (type 2) to MongoDB
Wrote 1 violations to MongoDB.

Writing batch 2 (type 2) to MongoDB
Wrote 2 violations to MongoDB.

Writing batch 3 (type 2) to MongoDB
Wrote 2 violations to MongoDB.

Writing batch 4 (type 2) to MongoDB
Wrote 1 violations to MongoDB.

Writing batch 5 (type 2) to MongoDB
Empty batch.

Writing batch 6 (type 2) to MongoDB
Wrote 1 violations to MongoDB.

Writing batch 7 (type 2) to MongoDB
Empty batch.

Writing batch 8 (type 2) to MongoDB
Wrote 7 violations to MongoDB.

Writing batch 9 (type 2) to MongoDB
Wrote 1 violations to MongoDB.

Writing batch 10 (type 2) to MongoDB
Wrote 1 violations to MongoDB.

Writing batch 11 (type 2) to MongoDB
Wrote 1 violations to MongoDB.

Writing batch 12 (type 2) to MongoDB
Empty batch.

Writing batch 13 (type 2) to MongoDB
Empty batch.

Writing batch 14 (type 2) to MongoDB
Wrote 5 violations to MongoDB.

Writing batch 15 (type 2) t

ERROR:root:KeyboardInterrupt while sending command.
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/site-packages/py4j/java_gateway.py", line 1038, in send_command
    response = connection.send_command(command)
  File "/opt/conda/lib/python3.8/site-packages/py4j/clientserver.py", line 511, in send_command
    answer = smart_decode(self.stream.readline()[:-1])
  File "/opt/conda/lib/python3.8/socket.py", line 669, in readinto
    return self._sock.recv_into(b)
KeyboardInterrupt
ERROR:py4j.clientserver:There was an exception while executing the Python Proxy on the Python Side.
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/site-packages/py4j/clientserver.py", line 617, in _call_proxy
    return_value = getattr(self.pool[obj_id], method)(*params)
  File "/opt/conda/lib/python3.8/site-packages/pyspark/sql/utils.py", line 272, in call
    raise e
  File "/opt/conda/lib/python3.8/site-packages/pyspark/sql/utils.py", line 269, in call
    self.func(Da

Interrupted by CTRL-C. Stopping all queries.


In [119]:
# testing 
after_count = violation_coll.count_documents({})
print(f"Violation count AFTER stream: {after_count}")
print(f"New records added: {after_count - before_count}")

Violation count AFTER stream: 50034
New records added: 34


In [155]:
rm -rf ./chk_violation_join_writer

In [ ]:
# testing 
violation_coll = db["violation"]
before_count = violation_coll.count_documents({})
print(f"Violation count BEFORE stream: {before_count}")

AB_join = AB_violation.writeStream \
    .foreachBatch(lambda df, batch_id: write_union_batch_to_mongo(df, batch_id, 1)) \
    .outputMode("append") \
    .option("checkpointLocation", "./chk_violation_join_writer") \
    .start()

try:
    AB_join.awaitTermination()
except KeyboardInterrupt:
    print("Stopping all ")
except Exception as e:
    print(e)
finally:
    AB_join.stop()

Violation count BEFORE stream: 50034

Writing batch 0 (type 1) to MongoDB
Empty batch.

Writing batch 1 (type 1) to MongoDB
{'event_id_1': 'f3162606-1b2e-407f-951d-61d14c0a7b09', 'batch_id_1': 1, 'speed_1': 125.2, 'event_time_1': datetime.datetime(2024, 1, 1, 8, 0, 2), 'producer_id_1': 1, 'event_id_2': 'ac84bb22-32be-4993-a655-49874adeecc4', 'batch_id_2': 4, 'speed_2': 117.6, 'event_time_2': datetime.datetime(2024, 1, 1, 8, 0, 31), 'producer_id_2': 2, 'car_plate': 'CJP 278', 'avg_speed': 124.13793103448276}
{'event_id_1': '3ca5e275-0cf6-43a0-8c28-65040c60b8b2', 'batch_id_1': 1, 'speed_1': 140.0, 'event_time_1': datetime.datetime(2024, 1, 1, 8, 0, 3), 'producer_id_1': 1, 'event_id_2': 'ee787113-eb0d-49d6-9cf9-054569625449', 'batch_id_2': 4, 'speed_2': 142.8, 'event_time_2': datetime.datetime(2024, 1, 1, 8, 0, 28), 'producer_id_2': 2, 'car_plate': 'HP 1', 'avg_speed': 144.0}
{'event_id_1': '7214629f-2fbb-4787-b953-e163720a6c61', 'batch_id_1': 1, 'speed_1': 133.3, 'event_time_1': datetime


Writing batch 4 (type 1) to MongoDB
{'event_id_1': '7e2c7380-bb0a-4237-a7b0-ada854450864', 'batch_id_1': 4, 'speed_1': 124.0, 'event_time_1': datetime.datetime(2024, 1, 1, 8, 19, 47), 'producer_id_1': 1, 'event_id_2': '7a0ed97c-8aeb-413b-831e-bf10dc172ada', 'batch_id_2': 37, 'speed_2': 118.8, 'event_time_2': datetime.datetime(2024, 1, 1, 8, 20, 15), 'producer_id_2': 2, 'car_plate': 'CI 9861', 'avg_speed': 128.57142857142858}
{'event_id_1': '2cedaf61-3ecb-4c24-8c7a-0aa7b3d88f7e', 'batch_id_1': 4, 'speed_1': 112.2, 'event_time_1': datetime.datetime(2024, 1, 1, 8, 19, 44), 'producer_id_1': 1, 'event_id_2': '1766270c-3293-4b56-9fef-8b16250fe1c3', 'batch_id_2': 37, 'speed_2': 106.1, 'event_time_2': datetime.datetime(2024, 1, 1, 8, 20, 15), 'producer_id_2': 2, 'car_plate': 'DOR 6', 'avg_speed': 116.12903225806451}
{'event_id_1': 'fb704780-229d-4d81-8afe-e766a61bf515', 'batch_id_1': 3, 'speed_1': 127.8, 'event_time_1': datetime.datetime(2024, 1, 1, 8, 13, 7), 'producer_id_1': 1, 'event_id_2'


Writing batch 6 (type 1) to MongoDB
{'event_id_1': 'd62f5e13-1055-40e7-86a8-ed8c697c4b84', 'batch_id_1': 6, 'speed_1': 149.1, 'event_time_1': datetime.datetime(2024, 1, 1, 8, 35, 23), 'producer_id_1': 1, 'event_id_2': '574ef9f7-5ca3-4e2d-a323-3f4530e0ce1d', 'batch_id_2': 55, 'speed_2': 153.0, 'event_time_2': datetime.datetime(2024, 1, 1, 8, 35, 46), 'producer_id_2': 2, 'car_plate': 'TB 5190', 'avg_speed': 156.52173913043478}
{'event_id_1': '261e1e49-07ad-49cc-b4a7-a418792be4a4', 'batch_id_1': 5, 'speed_1': 134.2, 'event_time_1': datetime.datetime(2024, 1, 1, 8, 25, 24), 'producer_id_1': 1, 'event_id_2': 'd01bcc6c-1c6e-4132-ab8c-aef2e2b91136', 'batch_id_2': 52, 'speed_2': 129.0, 'event_time_2': datetime.datetime(2024, 1, 1, 8, 25, 51), 'producer_id_2': 2, 'car_plate': 'ZBR 9', 'avg_speed': 133.33333333333334}
{'event_id_1': '654727a6-0002-46dd-ab19-8e468c001848', 'batch_id_1': 5, 'speed_1': 149.3, 'event_time_1': datetime.datetime(2024, 1, 1, 8, 25, 21), 'producer_id_1': 1, 'event_id_2

ERROR:root:KeyboardInterrupt while sending command.
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/site-packages/py4j/java_gateway.py", line 1038, in send_command
    response = connection.send_command(command)
  File "/opt/conda/lib/python3.8/site-packages/py4j/clientserver.py", line 511, in send_command
    answer = smart_decode(self.stream.readline()[:-1])
  File "/opt/conda/lib/python3.8/socket.py", line 669, in readinto
    return self._sock.recv_into(b)
KeyboardInterrupt


{'event_id_1': '8ed66d4a-c486-4f81-aba2-747ba57894e2', 'batch_id_1': 6, 'speed_1': 119.7, 'event_time_1': datetime.datetime(2024, 1, 1, 8, 35, 24), 'producer_id_1': 1, 'event_id_2': 'def44cb3-9553-4637-a9dc-5c33c6c2eab6', 'batch_id_2': 63, 'speed_2': 123.3, 'event_time_2': datetime.datetime(2024, 1, 1, 8, 35, 55), 'producer_id_2': 2, 'car_plate': 'GNU 052', 'avg_speed': 116.12903225806451}
{'event_id_1': 'dcc2f0f5-7de5-4563-b166-6967c54a0ad5', 'batch_id_1': 6, 'speed_1': 117.1, 'event_time_1': datetime.datetime(2024, 1, 1, 8, 35, 20), 'producer_id_1': 1, 'event_id_2': '9d03b365-b844-4566-80f0-37b662cf8d65', 'batch_id_2': 62, 'speed_2': 117.2, 'event_time_2': datetime.datetime(2024, 1, 1, 8, 35, 51), 'producer_id_2': 2, 'car_plate': 'HMU 040', 'avg_speed': 116.12903225806451}
{'event_id_1': '508db83a-e1cb-4d4b-9c7b-156573739994', 'batch_id_1': 6, 'speed_1': 114.7, 'event_time_1': datetime.datetime(2024, 1, 1, 8, 35, 20), 'producer_id_1': 1, 'event_id_2': '3c86113d-ee96-4f1a-bdd6-c296f4d

In [131]:
# testing 
after_count = violation_coll.count_documents({})
print(f"Violation count AFTER stream: {after_count}")
print(f"New records added: {after_count - before_count}")

Violation count AFTER stream: 50107
New records added: 73


In [156]:
rm -rf ./chk_violation_ab_writer

In [157]:
rm -rf ./chk_violation_bc_writer

In [158]:
rm -rf ./chk_violation_single_writer

In [ ]:
# AB Violation Stream Writer 
AB_join = AB_join.writeStream \
    .foreachBatch(lambda df, batch_id: write_union_batch_to_mongo(df, batch_id, 1)) \
    .outputMode("append") \
    .option("checkpointLocation", "./chk_violation_ab_writer") \
    .start()

# BC Violation Stream Writer
BC_join = BC_join.writeStream \
    .foreachBatch(lambda df, batch_id: write_union_batch_to_mongo(df, batch_id, 1)) \
    .outputMode("append") \
    .option("checkpointLocation", "./chk_violation_bc_writer") \
    .start()

# Single-point Violation Stream Writer
single_join = violations_union.writeStream \
    .foreachBatch(lambda df, batch_id: write_union_batch_to_mongo(df, batch_id, 2)) \
    .outputMode("append") \
    .option("checkpointLocation", "./chk_violation_single_writer") \
    .start()

# Terminate All Streams
try:
    AB_join.awaitTermination()
    BC_join.awaitTermination()
    single_join.awaitTermination()
except KeyboardInterrupt:
    print(" Interrupted by user. Stopping all streams.")
except Exception as e:
    print(f"Error: {e}")
finally:
    AB_join.stop()
    BC_join.stop()
    single_join.stop()
    print("All writers stopped.")


### Task 2.2 Data Visualisation

In this task, you will implement a program to visualize the joined streaming data. For the incoming camera event(s), 
* plot the number of violation against arrival time. You need to label some interesting points such as maximum and minimum values. 
* In addition to that, plot the speed against arrival time. You need to include some interesting points such as average and maximum values.

For visualization on the data stored in the database, you have to plot a map using camera location. On the map, annotate
* number of violations between the checkpoints
* identify hotspot (e.g. when number of violations exceed certain threshold within a time in a day)

Explain and justify the plots and the inclusion of the interesting points. Set your own threshold for the hotspot.

If you are running this task in a separate Jupyter notebook file, save the file as **xxx_assignment02_visualisation.ipynb**, where **xxx** represents the student IDs of the group members.

## <span style="color:#0b486b">Part 3: Documentation and comments to describe the proposed solution in the submitted notebook</span>

You should include sufficient comments and explanation Tasks 1 and 2 to describe your algorithm and/or code implementation. Please add additional markdown cells to explain your work. Adding extra illustrations to describe your method will also add to the marks in this part.

## <span style="color:#0b486b">Part 4: Code demo and interview</span>

In this task, you will present and showcase the simulation. After the assignment due date, you will be asked to attend an interview/demo session to showcase your application. Your interviewer will ask you a few questions in relation to your application and assess your understanding.

During the code demo, your work will be evaluated and assessed based on the marking guideline. Group members will obtain the same marks based on the code demo, unless there is an imbalance in contributions between students in a team. Additionally, each team member will be interviewed to explain the submitted work. The interview represents an individual assessment and a score between 0 and 1 will be awarded, which is then multipled with the marks obtained during the code demo.

Interviews for Assignment-2 will be conducted during Week 12 lab sessions. If you are granted an extension from special consideration, the interview will be conducted during SWOT-VAC week.